Newsletters: add 180 (2021-12-22) (#702)
bitcoinops/bitcoinops.github.io: _posts/en/newsletters/2021-12-22-newsletter.md

and other necessary information.  Onion messages and offers would
continue as draft specifications for the remainder of the year, but they
would receive additional development, including a proposal to use them
to [reduce the impact of stuck payments][offers stuck].

<!-- TODO:EXTRA: /en/newsletters/2021/01/06/#trampoline-routing -->

## February

Contributors to Bitcoin [advanced][safegcd blog] the state of research
into an improved signature creation and verification algorithm, and then
used their research to produce a variation with additional improvements.
When implemented in libsecp256k1 ([1][secp831], [2][secp906]) and
[later][bcc21573] in Bitcoin Core, this reduced the amount of time it
takes to verify signatures by about 10%---a significant improvement when
verifying the roughly billion signatures in Bitcoin's block chain.
Several cryptographers worked on ensuring the change was mathematically
sound and safe to use.  The change also provides a significant boost to
the speed of securely creating signatures on low-powered hardware
signing devices.

[Channel jamming attacks][topic channel jamming attacks], a known

> problem for LN since 2015, received continued discussion throughout the

Author: I think it wasn't "known" before then, and the description of it was before there actually was a LN (i.e. pre-segwit), so it wasn't exploitable originally.

year, with a [variety][jam1] of [possible][jam2] solutions
[discussed][jam3].  Unfortunately no widely accepted solution was found

> and the problem remained unmitigated by years end.

Reviewer: ```suggestion
and the problem remained unmitigated by year's end.
```

signatures] and [tapscript][topic tapscript] had been
[merged][bcc#19953] into Bitcoin Core.  This largely completed the work
of protocol developers; it was now up to the community to activate
taproot if they wished, and for wallet developers to begin adding
support for it and related technology like [bech32m][topic bech32]
addresses.

{% comment %}<!-- comments in bold text below tweak auto-anchors to prevent ID conflicts -->{% endcomment %}

- **January<!--taproot-->** began with the [release][bcc21] of Bitcoin Core 0.21.0,
  which was the first release to contain support for [signets][topic
  signet], including the default signet where taproot had been
  activated---giving users and developers an easy place to start
  testing.

- **February<!--taproot-->** saw the [first][tapa1] of [many][tapa2] scheduled
  [meetings][tapa3] in the `##taproot-activation` IRC channel, which
  would become the primary hub for discussion among developers, users,
  and miners for how to activate taproot.

- **March<!--taproot-->** began with many discussion participants tentatively agreeing
  to [try][speedy trial] a particular activation approach named *speedy

>   trial*, which was design to gather rapid feedback from miners but also

Reviewer 2: ```suggestion
  trial*, which was designed to gather rapid feedback from miners but also
```

    segwit version addresses fixes a bug which could've caused taproot
    users to lose money in some very rare cases.  (Original segwit
    outputs created from bech32 addresses are safe and unaffected by the
    bug.)

  {% comment %}
  /en/newsletters/2021/03/03/#rust-lightning-794
  /en/newsletters/2021/03/10/#documenting-the-intention-to-use-and-build-upon-taproot
  /en/newsletters/2021/03/24/#discussion-of-quantum-computer-attacks-on-taproot
  /en/newsletters/2021/03/24/#bitcoin-core-20861
  /en/newsletters/2021/03/31/#should-block-height-or-mtp-or-a-mixture-of-both-be-used-in-a-soft-fork-activation-mechanism
  {% endcomment %}

- **April<!--taproot-->** almost saw activation progress derail as protocol developers
  and some users [debated][tapa4] the merits of two slightly different versions
  of the speedy trial activation mechanism.  However, the authors of
  different implementations of the two different versions came to a
  [compromise][bcc#21377] that allowed a Bitcoin Core [version][bcctap]
  to be released with activation parameters.

<!-- TODO:EXTRA: BIPs drama? -->

- **May<!--taproot-->** was when miners [began][signal began] to be [able][signal

>   able] to signal their willingness to enforce taproot and a website for

Reviewer: Maybe:
```suggestion
  able] to signal their readiness to enforce taproot and a website for
```

> - **May<!--taproot-->** was when miners [began][signal began] to be [able][signal
>   able] to signal their willingness to enforce taproot and a website for

Reviewer 2: This read somewhat awkwardly
```suggestion
- **May<!--taproot-->** was when miners were [able][signal able] to [begin][signal began] signalling their willingness to enforce taproot and a website for
```

  [activity][rb#644] as taproot activation [approached][testing
  taproot].  Taproot's BIP was updated with [expanded test vectors][] to
  help wallet and infrastructure developers verify their
  implementations.

- **November<!--taproot-->** celebrated [taproot activation][].  There was some
  initial confusion as the miners of block {{site.trb}} and several
  subsequent blocks didn't include any taproot-spending transactions.
  However, thanks to the work of several developers and mining pool
  administrators, most blocks mined in subsequent days were ready to
  contain taproot-spending transactions.  Development and [testing][cbf
  verification] of taproot-ready software [continued][rb#691].

- **December<!--taproot-->** saw Bitcoin Core merge a PR that would
  allow [descriptor wallets][topic descriptors] to create
  [bech32m][topic bech32] addresses for receiving taproot payments.  LN
  devs also further discussed [making use][ln ptlcs] of taproot's
  features.

Despite complications choosing an activation mechanism for taproot and
some minor confusion immediately after its activation, the final steps
of adding support for the taproot soft fork to Bitcoin overall seemed to

> go well.  This is hardly the end the of the story for taproot.  Optech

Reviewer 2: ```suggestion
go well.  This is hardly the end of the story for taproot.  Optech
```

expects to continue to spend a significant amount of time writing about
it in the coming years as wallet and infrastructure developers begin
making use of its many features.

</div>

## April

> LND added [support][lnd#5709] in April for making Atomic Multipath

Reviewer 3: Suggest /en/topics/atomic-multipath/ topic

Payments (AMP), also called original AMPs due to being described earlier
than the [Simplified Multipath Payments][topic multipath payments]
(SMPs) all major LN implementations currently support.  AMPs have a
privacy advantage over SMPs and also ensure that the receiver has
received all parts before they claim the payment.  Their downside is
that they don't produce cryptographic proof of payment.  LND implemented
them for use with [spontaneous payments][topic spontaneous payments]
which fundamentally can't provide a proof of payment, eliminating from
consideration the one significant downside of AMPs.

## May

A discrepancy between the specification of [BIP125][] transaction
[replacement][topic rbf] and the implementation in Bitcoin Core was
[disclosed][bip125 discrep] in May.  This didn't put any bitcoins at
risk, as far as we know, but it did spawn several discussions about the
risks to contract protocols such as LN from unexpected transaction relay

> behavior.

Author: I made a slight edit here, but I wanted to try to understand your reason for suggesting this change.

Also in May, the C-Lightning project [merged][cl#4489] a plugin for
managing [dual-funded channels][topic dual funding]---channels where
both parties can provide some amount of the initial funding.  In
addition to prior dual-funding work [merged][cl#4410] this year, this
allows the party initiating the channel open to not only spend funds
through the channel but also receive them in the channel's initial
state.  Dual-funding is particularly useful for merchants whose primary

> use of LN is receiving payments instead of sending them.

Reviewer: It's not clear to me why Dual-funding would be helpful for users that primarily receive. Either way, the counterparty has to either open the channel or participate in opening. Perhaps this point could be elaborated.

Author: I'll be pushing an edit but I'm not 100% sure it addresses your concern---I'm confused about why you're confused---so let me know if it doesn't help.

  watchtowers] with [anchor outputs][topic anchor outputs] and added a
  new `psbt` wallet subcommand for working with [PSBTs][topic psbt].

- [HWI 2.0.0][] contained support for multisig on the BitBox02, improved
  documentation, and support for paying `OP_RETURN` outputs with a
  Trezor.

- [C-Lightning 0.10.0][] contained a number of enhancements to its API
  and experimental support for [dual funding][topic dual funding].

- [BTCPay Server 1.1.0][] included [Lightning Loop][news53 lightning
  loop] support, added [WebAuthN/FIDO2][fido2 website] as a two-factor
  authentication option, made various UI improvements, and marked a
  switch to using [semantic versioning][semantic versioning website] for
  version numbers moving forward.

- [Eclair 0.6.0][] contained several improvements that enhanced user
  security and privacy.  It also provided compatibility with future
  software that may use [bech32m][topic bech32] addresses.

- [LND 0.13.0-beta][] improved feerate management by making [anchor
  outputs][topic anchor outputs] the default commitment transaction
  format, added support for using a pruned Bitcoin full node, allowed

>   receiving and sending payments using Atomic MultiPath ([AMP][topic

Reviewer 3: Should this topic link be AMP or Multipath payments?

  multipath payments]), and increased LND's [PSBT][topic psbt]
  capabilities

- [Bitcoin Core 22.0][] included support for [I2P][topic anonymity

>   networks] connections, removed support for [version 2 Tor][topic

Reviewer 3: Suggest removing the 'version 2 Tor' link since we have it just prior.

Reviewer: I don't see the same link anywhere else. I think you may be referring to the version 2 address announcements in the Bitcoin Core 0.21.0 notes, which refer to the format how to announce node addresses, while the "version 2 Tor" here refers to Tor versions.

Author: @Xekyo I think @bitschmidty is referring to both the I2P and Tor links in the paragraph pointing to the same page.

@bitschmidty I'd prefer to link both of them.  I know it's a bit silly, but I don't want to play favorites, so I'm linking them both.

  anonymity networks] connections, and enhanced support for [hardware
  wallets][topic hwi].

- [BDK 0.12.0][] added the ability to store data using Sqlite.

- [LND 0.14.0][] included additional [eclipse attack][topic eclipse
  attacks] protection (see [Newsletter #164][news164 ping]), remote
  database support ([Newsletter #157][news157 db]), faster pathfinding
  ([Newsletter #170][news170 path]), improvements for users of Lightning
  Pool ([Newsletter #172][news172 pool]), and reusable [AMP][topic amp]
  invoices ([Newsletter #173][news173 amp]).

- [BDK 0.14.0][] simplified adding an `OP_RETURN` output to a
  transaction and contained improvements for sending payments to
  [bech32m][topic bech32] addresses for taproot.

</div>

## June

A new [analysis][csb] discussed in June described an alternative way for
miners to select which transactions they want to include in the blocks
they create.  The new method is predicted to slightly increase miner
revenue in the short term.  In the long-term, wallets aware of the new

> method will be able to collaborate when [CPFP fee bumping][topic cpfp]

Reviewer: Candidate set-based blockbuilding would first need to be adopted by some miners, and that would likely only happen after it gets implemented in Bitcoin Core, so this feels a bit too optimistic.

```suggestion
method would be able to collaborate when [CPFP fee bumping][topic cpfp]
```

transactions, increasing the effectiveness of that technique.

Another attempt to make fee bumping more effective was a [proposal][rbf
default] to make [transaction replacement by fee][topic rbf] (RBF) the

> default behavior in Bitcoin Core.  This could help resolve some issues

Author: This was a useful comment, definitely not pedanticism.  It's very useful to me to know how other people read/misread my stuff, so please keep this class of comment coming.

with fee bumping in multiparty protocols and also improve privacy by
allowing more transactions to use uniform settings.  Related to privacy,
a separate [proposal][nseq default] suggested that wallets creating
taproot spends should set a default nSequence value even when they
don't need the features offered by [BIP68][]'s consensus enforced
sequence values; this would allow transactions created by software that
does need to use BIP68 to blend in with more common transactions.
Neither proposal seemed to make much progress despite few significant
objections.

> June also saw the merge of [first PR][bcc#20833] in a

Reviewer 2: ```suggestion
June also saw the merge of the [first PR][bcc#20833] in a
```

[series][bcc#22642] attempting to implement mempool package acceptance
in Bitcoin Core, the first step towards package relay.  [Package
relay][topic package relay] will allow relay nodes and miners to treat
packages of related transactions as if they were a single transaction

> for fee purposes.  A package might contain a parent transaction with a

Reviewer: ```suggestion
for feerate purposes.  A package might contain a parent transaction with a
```

low feerate and a child transaction with a high feerate; the
profitability of mining the parent transaction would incentivize miners
to also mine the parent transaction.  Although package mining has been
[implemented][bitcoin core #7600] in Bitcoin Core since 2016, there has
so far been no way for nodes to relay transactions as packages, meaning
low-feerate parent transactions might not reach miners during
high-feerate periods even if they have high-feerate children.  This
makes [CPFP fee bumping][topic cpfp] unreliable for contract protocols
using presigned transactions, such as LN.  Package relay hopes to solve
this key safety issue.

> An idea original proposed in 2019 for LN saw renewed life in June.  The

Reviewer 2: ```suggestion
An idea originally proposed in 2019 for LN saw renewed life in June.  The
```

nodes the ability to buy incoming capacity that allows them to receive
instant payments. A node that sees the advertisement can simultaneously
pay for and receive the incoming capacity using a [dual funded][topic
dual funding] channel open. Although there’s no way to enforce that the
advertising node actually routes payments, the proposal does incorporate
an earlier proposal also [later used][lnd#5709] in Lightning Pool that
prevents the advertiser from using their money for other purposes until
the agreed upon lease period has concluded.  That means refusing to
route payments would provide no advantages but would deny the
advertising node the opportunity to earn routing fees.

<!-- TODO:EXTRA:table comparing liquidity ads and Pool -->

Three years after [first being proposed][descriptor gist] for Bitcoin
Core, [draft BIPs][descriptor bips1] were [created][descriptor bips2] for
[output script descriptors][topic descriptors].  Descriptors are strings
that contain all the information necessary to allow a wallet or other
program to track payments made to or spent from a particular script or
set of related scripts (i.e. an address or a set of related addresses

> such as in an HD wallet).  Descriptors combine well with [miniscript][topic miniscript] in

Reviewer 3: suggest linking to /en/topics/hd-key-generation/ topic

allowing a wallet to handle tracking and signing for a larger variety of
scripts.  They also combine well with [PSBTs][topic psbt] in allowing
the wallet to determine which keys it controls in a multisig script.  By
the end of the year, Bitcoin Core had made descriptor-based wallets its
[default][descriptor default] for newly-created wallets.

> A common way of opening LN channels that had never before been official

Reviewer 3: s/official/officially

Reviewer: Wasn't it "officially supported" by one LN implementation? Perhaps instead of "officially supported" it could be described as previously "not part of the LN specification".

channels where the funder gives some or all of their initial funds to
the acceptor. Those funds are not secure until the channel open
transaction receives a sufficient number of confirmations, so there’s no
risk to the acceptor spending some of those funds back through the
funder using the standard LN protocol.  For example, Alice has several
BTC in an account at Bob’s custodial exchange. Alice asks Bob to open a
new channel paying her 1.0 BTC. Because Bob trusts himself not to
double-spend the channel he just opened, he can allow Alice to send 0.1
BTC through his node to third-party Carol even before the channel open
transaction has received a single confirmation.  Specifying the behavior
should help improve interoperability between LN nodes and the merchants
who offer this service.

{:.center}
![Zero-conf channel illustration](/img/posts/2021-07-zeroconf-channels.png)

Two related proposals for new signature hash (sighash) types were
[combined][sighash combo] into [BIP118][].  `SIGHASH_NOINPUT`, proposed
in 2017 and partly based on previous proposals going back a decade, was
superseded by [`SIGHASH_ANYPREVOUT`][topic sighash_anyprevout] first

> [proposed][anyprevout proposed] in 2019.  The new sighash types will

Reviewer 3: I think mentioning `SIGHASH_ANYPREVOUTANYSCRIPT` in some capacity before "the new sighash typeS" would help with clarity.

allow offchain protocols such as LN and [vaults][topic vaults] to reduce

> the amount of intermediate states they need to retain, greatly reducing

Reviewer 2: nit: s/amount/number/ since it's a countable noun state

storage requirements and complexity.  For multiparty protocols, the
benefits may be even more significant by eliminating the number of
different states that need to be generated in the first place.

<!--
TODO:EXTRA LN migration to real databases:
  /en/newsletters/2021/07/14/#lnd-5447
  /en/newsletters/2021/07/21/#eclair-1871
  /en/newsletters/2021/10/06/#lnd-5366
-->

## August

> Fidelity bonds is an idea [proposed][wiki contract] at least as early as
> 2010 for locking up bitcoins for a period of time in order to create a
> cost for misbehavior.  Because the bitcoins can't be used again until
> their time lock expires, anyone banned or otherwise penalized during the

Reviewer 4: Perhaps it's worth distinguishing this as an already-implementable idea for applications, and not a proposal for a protocol change (which much of the other things called "proposal" in this newsletter are). I'm imagining a (not-well-informed) reader that gets confused thinking there's a proposal for banning users temporarily.
```suggestion
Since at least as early as 2010, fidelity bonds have been [discussed][wiki contract]
as a way for Bitcoin applications to penalize misbehavior by locking up
bitcoins for a period of time.  Because the bitcoins can't be used again until
their time lock expires, anyone banned or otherwise penalized during the
```

routing nodes only charged a percentage of the amount routed without
charging a minimum *base fee* on every payment.  Others felt
differently.  By the end of the year, a [variation][cl#4771] on the
technique would be implemented in C-Lightning.

<div markdown="1" class="callout" id="optech">
### 2021 summary<br>Bitcoin Optech

In Optech's fourth year, we published 51 weekly [newsletters][], added 30
new pages to our [topics index][], published a [contributed blog
post][additive batching], and wrote (with the help of [two][zmn guest]
guest [posters][darosior guest]) a 21-part series about [preparing for
taproot][p4tr]).  Altogether, Optech published over 80,000 words about
Bitcoin software research and development this year, the rough
equivalent of a 250-page book. <!-- wc -w _posts/en/newsletters/2021-*
_includes/specials/taproot/en/* -->

</div>

## September

One feature Bitcoin developers have long discussed is the ability to
send bitcoins to a script which could limit which other scripts could
later receive those bitcoins, a mechanism called [covenants][topic
covenants].  For example, Alice receives bitcoins to a script that can

> be spent by her hot wallet---but only by sending it to second script

Reviewer 3: s/to second/to a second

that time delays any further spend by her hot wallet.  During the time
delay, her cold wallet can claim the funds.  If it doesn't, and the time
delay passes, Alice's hot wallet can spend the funds freely.  In
September, a new `OP_TAPLEAF_UPDATE_VERIFY` opcode was
[proposed][op_tluv] for creating these sort of covenants in a way that
takes particular advantage of taproot's ability to spend funds either

> using just a signature (keypath spending) or a MAST-like tree of scripts

Reviewer 3: Suggest MAST link to /en/topics/mast/

(scriptpath spending).  The new opcode would be especially useful for
creating [joinpools][topic joinpools] that could significantly increase
privacy by allowing multiple users to easily and trustlessly share
ownership of a UTXO.

> 

Author: As much as possible, I've tried to mention each thing (and closely related things) just once, so I'd prefer to stick to the June mention of package relay and associated things.

[taproot][topic taproot] in LN, including [PTLCs][topic ptlc],
[MuSig2][topic musig] for [multisignatures][topic multisignature], and
[eltoo][topic eltoo]; moving specification discussion from IRC to video
chats; changes to the current BOLTs specification model; [onion
messages][topic onion messages] and [offers][topic offers]; [stuckless
payments][]; [channel jamming attacks][topic channel jamming attacks]
and various mitigations; and [trampoline routing][topic trampoline
payments].

## December

For single-sig onchain transactions, fee bumping a transaction's feerate
to encourage miners to confirm it sooner is a relatively straightforward
operation.  But for contract protocols such as LN and [vaults][topic
vaults], not all the signers who authorized a spend may be available
when fee bumping is needed.  Worse, contract protocols often require
certain transactions to be confirmed by a deadline---or an honest user
could lose money.  December saw the [publication][fee bump research] of
research related to choosing effective fee bumping mechanisms for
contract protocols, helping spur discussion of solutions to this
important long-term problem.

> 

Author: Gonna leave this out.  I don't think the merge itself is that big of a deal and we do already have a lot of taproot/post-taproot coverage this year.  Also, I'm lazy.

---
title: 'Bitcoin Optech Newsletter #180: 2021 Year-in-Review Special'
permalink: /en/newsletters/2021/12/22/
name: 2021-12-22-newsletter
slug: 2021-12-22-newsletter
type: newsletter
layout: newsletter
lang: en

excerpt: >
  This special edition of the Optech Newsletter summarizes notable
  developments in Bitcoin during all of 2021.
---

{{page.excerpt}} It's the sequel to our summaries from [2018][2018
summary], [2019][2019 summary], and [2020][2020 summary].

## Contents

* January
  * [Signet in Bitcoin Core](#signet)
* FIXME:fill out table of contents

## January

{:#signet}
After years of discussion, January saw the first [release][bcc21] of a
Bitcoin Core version supporting [signets][topic signet], following prior
[support][cl#2816] by C-Lightning and followed by [support][lnd#5025] in

> LND.  Signets are test networks anyone can used to simulate Bitcoin's

Reviewer 2: ```suggestion
LND.  Signets are test networks anyone can be used to simulate Bitcoin's
```
